astropy/units/tests/test_physical.py

```diff
@@ -1,60 +1,436 @@
+    (u.N, u.J),
+    (u.barn, u.imperial.deg_F),
+]
+
+
+@pytest.mark.parametrize("unit1, unit2", nonequivalent_unit_pairs)
+def test_physical_type_instance_inequality(unit1, unit2):
+    """
+    Test that `physical.PhysicalType` instances for units with different
+    dimensionality are considered unequal.
+    """
+    physical_type1 = physical.PhysicalType(unit1, "ptype1")
+    physical_type2 = physical.PhysicalType(unit2, "ptype2")
+    assert (physical_type1 != physical_type2) is True
+    assert (physical_type1 == physical_type2) is False
+
+
+physical_type_with_expected_str = [
+    (length, "length"),
+    (speed, "speed"),
+    (pressure, "{'energy density', 'pressure', 'stress'}"),
+    (u.deg_C.physical_type, "temperature"),
+    ((u.J / u.K / u.kg).physical_type, "{'specific entropy', 'specific heat capacity'}"),
+]
+
+physical_type_with_expected_repr = [
+    (length, "PhysicalType('length')"),
+    (speed, "PhysicalType('speed')"),
+    (pressure, "PhysicalType({'energy density', 'pressure', 'stress'})"),
+    (u.deg_C.physical_type, "PhysicalType('temperature')"),
+    ((u.J / u.K / u.kg).physical_type,
+     "PhysicalType({'specific entropy', 'specific heat capacity'})"),
+]
+
+
+@pytest.mark.parametrize("physical_type, expected_str", physical_type_with_expected_str)
+def test_physical_type_str(physical_type, expected_str):
+    """Test using `str` on a `PhysicalType` instance."""
+    assert str(physical_type) == expected_str
+
 
+@pytest.mark.parametrize(
+    "physical_type, expected_repr", physical_type_with_expected_repr
+)
+def physical_type_repr(physical_type, expected_repr):
+    """Test using `repr` on a `PhysicalType` instance."""
+    assert repr(physical_type) == expected_repr
 
-def test_data_quantity():
-    assert u.byte.physical_type == 'data quantity'
-    assert u.bit.physical_type == 'data quantity'
+
+def test_physical_type_hash():
+    """Test that a `PhysicalType` instance can be used as a dict key."""
+    dictionary = {length: 42}
+    assert dictionary[length] == 42
+
+
+def test_unrecognized_unit_physical_type():
+    """
+    Test basic functionality for the physical type of an unrecognized
+    unit.
+    """
+    unrecognized_unit = u.Unit("parrot", parse_strict="silent")
+    physical_type = unrecognized_unit.physical_type
+    assert isinstance(physical_type, physical.PhysicalType)
+    assert physical_type == "unknown"
+
+
+invalid_inputs = [(42,), ("valid input", 42)]
+
+
+@pytest.mark.parametrize("invalid_input", invalid_inputs)
+def test_invalid_physical_types(invalid_input):
+    """
+    Test that `PhysicalType` cannot be instantiated when one of the
+    supplied names is not a string, while making sure that the physical
+    type for the unit remains unknown.
+    """
+    obscure_unit = u.s ** 87
+    with pytest.raises(ValueError):
+        physical.PhysicalType(obscure_unit, invalid_input)
+    assert obscure_unit.physical_type == "unknown"
+
+
+def _undef_physical_type(unit):
```

Comment:
I think the way to do this is with `teardown_function` - that way it is guaranteed to run also if the test fails.

(I think all this is easier with a test class as suggested in https://github.com/astropy/astropy/pull/11204/files#r558759530 below)